File: fortran90/day5_part1_2018.f90
program day5
  implicit none
  character(len=100000) :: polymer
  integer :: i, j, n
  logical :: reacted

  open(unit=10, file='input.txt', status='old')
  read(10, '(A)') polymer
  close(10)

  n = len_trim(polymer)
  do while (.true.)
    reacted = .false.
    do i = 1, n-1
      if (iachar(polymer(i:i)) /= iachar(polymer(i+1:i+1)) + 32 .and. &
           iachar(polymer(i:i)) /= iachar(polymer(i+1:i+1)) - 32) cycle
      polymer = polymer(1:i-1) // polymer(i+2:n)
      n = n - 2
      reacted = .true.
    end do
    if (.not. reacted) exit
  end do

  print *, 'Units remaining:', n
end program day5

File: zig/day5_part1_2018.zig

const std = @import("std");
const fs = std.fs;
const io = std.io;
const mem = std.mem;

pub fn main() !void {
    // Open the input file
    const file = try fs.cwd().openFile("input.txt", .{});
    defer file.close();

    // Read the entire file content
    const content = try file.readToEndAlloc(std.heap.page_allocator, 1024 * 1024);
    defer std.heap.page_allocator.free(content);

    // Create a buffer to store the reduced polymer
    var buffer = try std.ArrayList(u8).initCapacity(std.heap.page_allocator, content.len);
    defer buffer.deinit();

    // Process the polymer
    for (content) |unit| {
        if (buffer.items.len > 0 and areReactive(buffer.items[buffer.items.len - 1], unit)) {
            _ = buffer.pop();
        } else {
            try buffer.append(unit);
        }
    }

    // Print the result
    const stdout = io.getStdOut().writer();
    try stdout.print("Remaining units: {}\n", .{buffer.items.len});
}

fn areReactive(a: u8, b: u8) bool {
    return (a ^ b) == 32;
}
